File: command/python/invoker.py

class Invoker:
    def __init__(self):
        self._commands = []
        self._undone_commands = []

    def execute_command(self, command):
        command.execute()
        self._commands.append(command)

    def undo(self):
        if self._commands:
            command = self._commands.pop()
            command.undo()
            self._undone_commands.append(command)

    def redo(self):
        if self._undone_commands:
            command = self._undone_commands.pop()
            command.execute()
            self._commands.append(command)

File: command/java/Invoker.java
package command.java;

import java.util.ArrayList;
import java.util.List;

import command.java.interfaces.Command;

public class Invoker {
    private List<Command> commands = new ArrayList<>();
    private List<Command> undoneCommands = new ArrayList<>();

    public void executeCommand(Command command) {
        command.execute();
        commands.add(command);
    }

    public void undo() {
        if (!commands.isEmpty()) {
            Command command = commands.remove(commands.size() - 1);
            command.undo();
            undoneCommands.add(command);
        }
    }

    public void redo() {
        if (!undoneCommands.isEmpty()) {
            Command command = undoneCommands.remove(undoneCommands.size() - 1);
            command.execute();
            commands.add(command);
        }
    }
}
